readme.md
## webpack-compass-imagehelper


this is copied from https://github.com/phlppschrr/gulp-compass-imagehelper and adjusted to work with webpack instead of gulp, it lets
you use the compass image helper functions together with a webpack/sass-loader enviroment.


### plugin options

```javascript
const CompassImageHelperPlugin = require('webpack-compass-imagehelper');

...

new CompassImageHelperPlugin({
			targetFile: 'target/_images.scss',
			images_path: 'src/main/scss/',
			pattern: 'resources/**/*',
			css_path: 'images/',
			sizeLimit: 10240, // size limit for image embedding
			prefix: 'icon--'
		})
```


This plugin generates a helper .scss file, which you have to @import into your own sass project. 
The generated sass file acts as a polyfill: Inside the generated file is a sass map which contains all 
the image infos including a inlined data version. The default mustache template also outputs %placeholders for 
each image. Feel free to modify the template to your needs. Additional there are the following helper function which mimic 
the native functions from Compass:

### Supported Compass functions
* **inline-image($image):** Embeds the contents of an image directly inside your stylesheet. All images are currently base64 encoded, in a future version SVG images will be UTF-8 encoded.
* **image-width($image):** Returns the width of the image found at the path supplied by $image. Warning, some SVG images may fail here and return `null`.
* **image-height($image)** Returns the height of the image found at the path supplied by $image. Warning, some SVG images may fail here and return `null`.
* **image-url($image, $only-path: false, $cache-buster: false)**  
  Generates a path to an asset found relative to the project's images directory.
  Passing a true value as the second argument will cause only the path to be returned instead of a url() function
  The third argument is used to control the cache buster on a per-use basis. When set to false no cache buster will be used. When true a md5-hash of the file is appended to the url. When a string, that value will be used as the cache buster. Be sure to set `images_path` and `css_path` options.
* **image-exists($image):** Returns true if the image exists in our helper file. (Not part of compass, but still its here)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v0.3.3](https://github.com/nick-less/webpack-compass-imagehelper/compare/v0.3.2...v0.3.3)

### Merged

- Bump gh-release from 6.0.3 to 6.0.4 [`#53`](https://github.com/nick-less/webpack-compass-imagehelper/pull/53)
- Bump glob from 7.2.0 to 8.0.1 [`#52`](https://github.com/nick-less/webpack-compass-imagehelper/pull/52)
- Bump webpack from 5.71.0 to 5.72.0 [`#51`](https://github.com/nick-less/webpack-compass-imagehelper/pull/51)
- Bump webpack from 5.70.0 to 5.71.0 [`#50`](https://github.com/nick-less/webpack-compass-imagehelper/pull/50)
- Bump minimist from 1.2.5 to 1.2.6 [`#49`](https://github.com/nick-less/webpack-compass-imagehelper/pull/49)
- Bump gh-release from 6.0.2 to 6.0.3 [`#48`](https://github.com/nick-less/webpack-compass-imagehelper/pull/48)
- Bump gh-release from 6.0.1 to 6.0.2 [`#47`](https://github.com/nick-less/webpack-compass-imagehelper/pull/47)
- Bump mocha from 9.2.1 to 9.2.2 [`#46`](https://github.com/nick-less/webpack-compass-imagehelper/pull/46)
- Bump webpack from 5.69.1 to 5.70.0 [`#45`](https://github.com/nick-less/webpack-compass-imagehelper/pull/45)
- Bump mocha from 9.2.0 to 9.2.1 [`#44`](https://github.com/nick-less/webpack-compass-imagehelper/pull/44)
- Bump webpack from 5.69.0 to 5.69.1 [`#43`](https://github.com/nick-less/webpack-compass-imagehelper/pull/43)
- Bump webpack from 5.68.0 to 5.69.0 [`#42`](https://github.com/nick-less/webpack-compass-imagehelper/pull/42)
- Bump auto-changelog from 2.3.0 to 2.4.0 [`#41`](https://github.com/nick-less/webpack-compass-imagehelper/pull/41)
- Bump webpack from 5.67.0 to 5.68.0 [`#40`](https://github.com/nick-less/webpack-compass-imagehelper/pull/40)

## [v0.3.2](https://github.com/nick-less/webpack-compass-imagehelper/compare/v0.3.1...v0.3.2) - 2022-01-30

### Merged

- Bump simple-get from 4.0.0 to 4.0.1 [`#39`](https://github.com/nick-less/webpack-compass-imagehelper/pull/39)
- Bump mocha from 9.1.4 to 9.2.0 [`#38`](https://github.com/nick-less/webpack-compass-imagehelper/pull/38)
- Bump node-fetch from 2.6.5 to 2.6.7 [`#37`](https://github.com/nick-less/webpack-compass-imagehelper/pull/37)
- Bump webpack from 5.66.0 to 5.67.0 [`#36`](https://github.com/nick-less/webpack-compass-imagehelper/pull/36)
- Bump mocha from 9.1.3 to 9.1.4 [`#35`](https://github.com/nick-less/webpack-compass-imagehelper/pull/35)
- Bump shelljs from 0.8.4 to 0.8.5 [`#34`](https://github.com/nick-less/webpack-compass-imagehelper/pull/34)
- Bump webpack from 5.65.0 to 5.66.0 [`#33`](https://github.com/nick-less/webpack-compass-imagehelper/pull/33)
- Bump image-size from 1.0.0 to 1.0.1 [`#32`](https://github.com/nick-less/webpack-compass-imagehelper/pull/32)
- Bump webpack from 5.64.4 to 5.65.0 [`#31`](https://github.com/nick-less/webpack-compass-imagehelper/pull/31)
- Bump webpack from 5.64.3 to 5.64.4 [`#30`](https://github.com/nick-less/webpack-compass-imagehelper/pull/30)
- Bump webpack from 5.64.2 to 5.64.3 [`#29`](https://github.com/nick-less/webpack-compass-imagehelper/pull/29)
- Bump webpack from 5.64.1 to 5.64.2 [`#28`](https://github.com/nick-less/webpack-compass-imagehelper/pull/28)
- Bump webpack from 5.64.0 to 5.64.1 [`#27`](https://github.com/nick-less/webpack-compass-imagehelper/pull/27)
- Bump webpack from 5.63.0 to 5.64.0 [`#26`](https://github.com/nick-less/webpack-compass-imagehelper/pull/26)
- Bump webpack from 5.62.1 to 5.63.0 [`#25`](https://github.com/nick-less/webpack-compass-imagehelper/pull/25)
- Bump webpack from 5.61.0 to 5.62.1 [`#24`](https://github.com/nick-less/webpack-compass-imagehelper/pull/24)
- Bump mime from 2.5.2 to 3.0.0 [`#23`](https://github.com/nick-less/webpack-compass-imagehelper/pull/23)
- Bump webpack from 5.60.0 to 5.61.0 [`#22`](https://github.com/nick-less/webpack-compass-imagehelper/pull/22)

## [v0.3.1](https://github.com/nick-less/webpack-compass-imagehelper/compare/v0.3.0...v0.3.1) - 2021-10-26

### Merged

- Bump webpack from 5.59.1 to 5.60.0 [`#21`](https://github.com/nick-less/webpack-compass-imagehelper/pull/21)
- Bump webpack from 5.59.0 to 5.59.1 [`#20`](https://github.com/nick-less/webpack-compass-imagehelper/pull/20)
- Bump webpack from 5.58.2 to 5.59.0 [`#19`](https://github.com/nick-less/webpack-compass-imagehelper/pull/19)
- Bump mocha from 9.1.2 to 9.1.3 [`#18`](https://github.com/nick-less/webpack-compass-imagehelper/pull/18)
- Bump webpack from 5.58.1 to 5.58.2 [`#17`](https://github.com/nick-less/webpack-compass-imagehelper/pull/17)

## [v0.3.0](https://github.com/nick-less/webpack-compass-imagehelper/compare/v0.2.0...v0.3.0) - 2021-10-12

### Merged

- Bump webpack from 4.46.0 to 5.58.1 [`#16`](https://github.com/nick-less/webpack-compass-imagehelper/pull/16)

## [v0.2.0](https://github.com/nick-less/webpack-compass-imagehelper/compare/v0.1.1...v0.2.0) - 2021-10-10

### Commits

- added simple caching [`e64b5aa`](https://github.com/nick-less/webpack-compass-imagehelper/commit/e64b5aadd8447677f061b22966cc3378162ebdd0)
- bump mocha, md5, mime [`d8aa1ab`](https://github.com/nick-less/webpack-compass-imagehelper/commit/d8aa1abf98b5f7466aa218987be15db3033cebec)

## v0.1.1 - 2021-10-10

### Merged

- Bump auto-changelog from 1.16.4 to 2.3.0 [`#14`](https://github.com/nick-less/webpack-compass-imagehelper/pull/14)
- Bump gh-release from 3.5.0 to 6.0.1 [`#15`](https://github.com/nick-less/webpack-compass-imagehelper/pull/15)
- Bump glob from 7.1.7 to 7.2.0 [`#13`](https://github.com/nick-less/webpack-compass-imagehelper/pull/13)
- Bump mocha from 8.4.0 to 9.0.0 [`#12`](https://github.com/nick-less/webpack-compass-imagehelper/pull/12)
- Bump image-size from 0.9.7 to 1.0.0 [`#11`](https://github.com/nick-less/webpack-compass-imagehelper/pull/11)

test/testPluginRun.js
var assert = require('assert');
var webpack = require('webpack');
var fs = require('fs');
var options = require('./webpack.config.js');

describe('Array', function() {
  describe('#runPlugin()', function() {
    webpack(options, function(err, stats) {
      if (err) {
        console.log(err)
      }
    })
    var size = 8946;
    it('should generate a file ', function() {
      var fstats = fs.statSync('./dist/_images.scss')
      assert.notEqual(null, fstats)
    });
    it('should generate a file of ' + size + ' bytes', function() {
      var fstats = fs.statSync('./dist/_images.scss')
      assert.equal(size, fstats.size)
    });
  });
});


test/webpack.config.js
const path = require('path');
const mode = process.env.NODE_ENV !== 'production'
const CompassImageHelperPlugin = require('../index');

module.exports = {
  entry:
      ['./test/entry.js'],
  output: {
    filename: 'out.js',
    path: path.resolve(__dirname, '../dist'),
    publicPath: '/'
  },
  plugins: [
  new CompassImageHelperPlugin({
    targetFile: './dist/_images.scss',
    images_path: './test/images/',
    pattern: '**/*',
    css_path: 'images/',
    sizeLimit: 10240, // size limit for image embedding
    prefix: 'icon--'
  })
],
  module: {
    rules: [
  
    ]
  }
};

index.js
var fs = require('fs');
var write = require('write');
var path = require('canonical-path');
var mustache = require('mustache');
var sizeOf = require('image-size');
var mime = require('mime');
var md5 = require('md5');
var appRoot = require('app-root-path').path;
var glob = require("glob");

function createImageAssetFile (options) {
		var images = [];
		  //var images_path = options.images_path || '';

		  if (!options) options = {};

		  if (!options.template)
		    options.template = path.join(__dirname, 'compass-imagehelper.mustache');

		  if (!options.targetFile)
		    options.targetFile = '_compass-imagehelper.scss';

		  if (!options.prefix)
		    options.prefix = '';
		    
			if (!options.sizeLimit)
		    options.sizeLimit = 10240;


		  var template = fs.readFileSync(options.template).toString();



		  var pathPrefix = function () {
		    var result = '';
		    if (options.http_images_path) {
		      result = options.http_images_path;
		    } else if (options.css_path && options.images_path) {
		      // relative path from css folder to images
		      result = path.relative(options.css_path, options.images_path);
		    } else if (options.images_path) {
		      // relative from project url
		      result = path.relative(appRoot, options.images_path);
		    }

		    // make sure pathPrefix ends with a trailing slash
		    if (result && result.substr(-1) != '/') {
		      result = result + '/';
		    }

		    return result;
		  };

		  var bufferContents = function (file) {
		  
		    if (!options.images_path) {
		      // autodetect images_path with the first file
		      options.images_path = path.relative(appRoot, file.base);
		    }

		    var imageInfo = {};
		    var data;
		    var encoding;

		    var mimetype = mime.getType(file.path);
		    var dimensions;

		    if (mimetype == 'image/svg+xml') {
		      data = file.contents.toString()
		          .replace(/"/g, '\'')
		          .replace(/%/g, '%25')
		          .replace(/</g, '%3C')
		          .replace(/>/g, '%3E')
		          .replace(/&/g, '%26')
		          .replace(/#/g, '%23')
		          .replace(/\s+/g, ' ')
		          .trim();
		      encoding = 'charset=US-ASCII';
		      dimensions = sizeOf(file.path);
		    } else {
		      data = file.contents.toString('base64');
		      encoding = 'base64';
		      dimensions = sizeOf(file.path);
		    }

		    imageInfo.width = dimensions.width;
		    imageInfo.height = dimensions.height;
		    imageInfo.mime = mimetype;
		    imageInfo.filename = path.basename(file.path);
		    imageInfo.basename = path.basename(file.path, path.extname(file.path));
		    imageInfo.dirname = path.basename(path.dirname(file.path));
		    imageInfo.ext = path.extname(file.path);
		    imageInfo.path = path.relative(options.images_path, file.path);
		    imageInfo.fullname = imageInfo.path.replace(/[\/\\\.]/g, '-'); // Replace /, \, and . with -
		    imageInfo.fullname = imageInfo.fullname.replace(/@/g, '\\@'); // Escape @
		    if (!isNaN(parseInt(imageInfo.fullname, 10))) {
		      // classnames must not start with an unescaped number.
		      // Instead of escaping we simply prefix starting numbers by '_'
		      imageInfo.fullname = '_' + imageInfo.fullname;
		    }

		    imageInfo.hash = md5(file.contents);
		    imageInfo.data = 'url("data:' + mimetype + ';' + encoding + ',' + data + '")';
		    images.push(imageInfo);

		  };

 		glob(options.images_path+options.pattern, function (er, files) {
			var changed =  true
			var newestFile = 0;
			var outfileTime = 0;
			if (fs.existsSync(options.targetFile)) {
					var outputFs  = fs.statSync( options.targetFile);
					var templateFs = fs.statSync(options.template);
					changed =  templateFs.mtime > outputFs.mtime || 0
					outfileTime = outputFs.mtime;
				}
				if (!changed) {
					for (var i = 0; i < files.length; i++) {
						var stats = fs.statSync(files[i]);
						if (newestFile < stats.mtime) {
							newestFile = stats.mtime;
						}
					}
					changed = newestFile > outfileTime
				}
	
				if (changed) {
					for (var i = 0; i < files.length; i++) {
  				
						var content = '';
						var stats = fs.statSync(files[i]);
						if (stats.size<options.sizeLimit) {
							content = fs.readFileSync(files[i]);
						}
					  
					  
						 bufferContents({path:files[i],contents:content, base:''})
					}
				   var fullPath =  options.targetFile;
					write.sync(fullPath, 
						Buffer.from(mustache.render(template, {
						prefix: options.prefix,
						path_prefix: pathPrefix(),
						items: images,
				  })));
				  console.log("webpack-compass-imagehelper: "+fullPath+" written for "+files.length+" images");
				} else {
					console.log('webpack-compass-imagehelper: no changes in ' + files.length + ' image files')
				}
			})

	}
	

function CompassImageHelperPlugin(options) {
	createImageAssetFile (options);

}

CompassImageHelperPlugin.prototype.apply = function(compiler) {

	/*
	compiler.plugin("compilation", function(compilation) {
	}); 
	compiler.plugin('emit', (compilation, done) => {
	    for (const filename in compilation.assets) {
	      const hash = crypto.createHash("md5");
	      hash.update(compilation.assets[filename].source());
	      const md5 = hash.digest('hex');

	      // Fingerprint .md5 file
	      compilation.assets[`${filename}.md5`] = {
	        source: () => md5,
	        size: () => md5.length,
	      };

	      // Identical to original file but with hash prepended.
	      compilation.assets[`${md5}-${filename}`] = compilation.assets[filename];
	    };
	    done();
	});
	*/
	
};

module.exports = CompassImageHelperPlugin;
